internal/util/pool_test.go

```diff
@@ -57,82 +63,114 @@ func TestRunParallelPoolMapBase(test *testing.T) {
+		if !reflect.DeepEqual(expected, output) {
+			test.Errorf("Case %d: Result not as expected. Expected: '%s', actual: '%s'.",
+				i, MustToJSONIndent(expected), MustToJSONIndent(output))
+			continue
 		}
 
 		// Check for the thread count last (this gives the workers a small bit of extra time to exit).
 		// Note that there may be other tests with stray threads, so we are allowed to have less than when we started.
-		time.Sleep(25 * time.Millisecond)
+		// The final thread that signals the output.IsDone() may not have enough time to exit,
+		// so we are allowed to end with one more thread than when we started.
 		endThreadCount := runtime.NumGoroutine()
-		if startThreadCount < endThreadCount {
-			test.Errorf("Case %d: Ended with more threads than we started with. Start: %d, End: %d.", i, startThreadCount, endThreadCount)
+		allowedEndThreadCount := endThreadCount - 1
+		if startThreadCount < allowedEndThreadCount {
+			test.Errorf("Case %d: Ended with more threads than we started with. Start: %d, End: %d.",
+				i, startThreadCount, allowedEndThreadCount)
 			continue
 		}
 	}
-}
 
-func TestRunParallelPoolMapCancel(test *testing.T) {
-	testCases := []struct {
-		numThreads int
-	}{
-		{1},
-		{2},
-		{3},
-		{4},
-		{10},
+	// Note that there may be other tests with stray threads, so we are allowed to have less than when we started.
+	// The final thread that signals the output.IsDone() may not have enough time to exit,
+	// so we are allowed to end with one more thread than when we started.
+	overallEndThreadCount := runtime.NumGoroutine()
+	allowedOverallEndThreadCount := overallEndThreadCount - 1
+	if overallStartThreadCount < allowedOverallEndThreadCount {
+		test.Fatalf("Ended with more threads than we started with. Start: %d, End: %d.",
+			overallStartThreadCount, allowedOverallEndThreadCount)
 	}
+}
 
+func TestRunParallelPoolMapCancel(test *testing.T) {
 	input := []string{
 		"A",
 		"BB",
 		"CCC",
 	}
 
-	for i, testCase := range testCases {
-		// Block until the first worker has started.
-		workWaitGroup := sync.WaitGroup{}
-		workWaitGroup.Add(1)
-
-		workFunc := func(input string) (int, error) {
-			// Signal on the first piece of work that we can make sure the workers have started up before we cancel.
-			if input == "A" {
-				workWaitGroup.Done()
-			}
+	// Count the number of active threads before running.
+	startThreadCount := runtime.NumGoroutine()
 
-			// Sleep for a really long time (for a test).
-			time.Sleep(1 * time.Hour)
+	// Block until the first worker has started.
+	workWaitGroup := sync.WaitGroup{}
+	workWaitGroup.Add(1)
 
-			return len(input), nil
+	// Block until the cancellation is called.
+	cancelWaitGroup := sync.WaitGroup{}
+	cancelWaitGroup.Add(1)
+
+	// Work will not start on input 'CCC'.
+	workFunc := func(input string) (int, error) {
+		// Signal on the second piece of work so that work on input 'A' will complete normally.
+		if input == "BB" {
+			workWaitGroup.Done()
```

Comment:
Instead of this, can we just cancel ourselves?
(Like just make the cancel context earlier and call `cancelFunc()`?